sql/core/src/main/scala/org/apache/spark/sql/execution/objects.scala

```diff
@@ -219,7 +219,33 @@ case class MapElementsExec(
   override def doConsume(ctx: CodegenContext, input: Seq[ExprCode], row: ExprCode): String = {
     val (funcClass, methodName) = func match {
       case m: MapFunction[_, _] => classOf[MapFunction[_, _]] -> "call"
-      case _ => classOf[Any => Any] -> "apply"
+      case _ => (child.output(0).dataType, outputObjAttr.dataType) match {
+        // if a pair of an argument and return types is one of specific types
+        // whose specialized method (apply$mc..$sp) is generated by scalac,
+        // Catalyst generated a direct method call to the specialized method.
+        // The followings are references for this specialization:
+        //   https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/transform/
+        //     SpecializeTypes.scala
+        //   http://www.cakesolutions.net/teamblogs/scala-dissection-functions
+        //   http://axel22.github.io/2013/11/03/specialization-quirks.html
+        case (IntegerType, IntegerType) => classOf[Int => Int] -> "apply$mcII$sp"
+        case (IntegerType, LongType) => classOf[Int => Long] -> "apply$mcJI$sp"
```

Comment:
is it possible do it in a composable way instead of enumerating all combinations?